tests/unit/cpu_profiler_test.cc

```diff
@@ -0,0 +1,174 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2023 ScyllaDB Ltd.
+ */
+
+#include <boost/test/tools/old/interface.hpp>
+#include <cstddef>
+#include <seastar/core/internal/cpu_profiler.hh>
+#include <seastar/core/internal/stall_detector.hh>
+#include <seastar/core/reactor.hh>
+#include <seastar/core/thread_cputime_clock.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/util/later.hh>
+#include <seastar/testing/test_case.hh>
+#include <seastar/testing/thread_test_case.hh>
+#include <atomic>
+#include <chrono>
+#include <sys/mman.h>
+
+#include "stall_detector_test_utilities.hh"
+
+struct temporary_profiler_settings {
+    std::chrono::nanoseconds prev_ns;
+    bool prev_enabled;
+
+    temporary_profiler_settings(bool enable, std::chrono::nanoseconds ns) {
+        prev_ns = engine().get_cpu_profiler_period();
+        prev_enabled = engine().get_cpu_profiler_enabled();
+
+        engine().set_cpu_profiler_period(ns);
+        engine().set_cpu_profiler_enabled(enable);
+    }
+
+    ~temporary_profiler_settings() {
+        engine().set_cpu_profiler_period(prev_ns);
+        engine().set_cpu_profiler_enabled(prev_enabled);
+    }
+};
+
+// If we set a timer to fire in N ms we can expect it to fire between N and (N + E) ms from
+// when it was set. Where E is some fixed error that arises from how granular a given timer
+// is. Therefore over a given period of time, M, if the timer is continously reset every
+// time it fires we can expect the timer to fire between M/N and M/(N+E) times.
+//
+// The function below takes this error into account and allows the actual samples taken
+// to be slightly less than the expect number of samples if there was no error.
+bool close_to_expected(size_t actual_size, size_t expected_size, double allowed_dev = 0.15) {
```

Comment:
This smells like setting up a kind of test which can flap. 

Did you try running this a like 100k times to see if it always succeeds?

Ideally the majority of the tests aren't timing dependent: i.e., they use a mock clock under the control of the test, but I don't think we should change that now.

Reply:
Agreed, I suppose a could write a timer class that could allow the test to control the clock and "fire" sampling events. At that point though we wouldn't be testing much.

Reply:
I disagree but we can move on.